[css-content] Fix issues with CSS `content` computed value serialization

nt1m opened the pull request:
#### c2222b448f74ddeb8f97fe41999e243a10235dc1
<pre>
[css-content] Fix issues with CSS `content` computed value serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=266828">https://bugs.webkit.org/show_bug.cgi?id=266828</a>
<a href="https://rdar.apple.com/120061551">rdar://120061551</a>

Reviewed by Darin Adler.

There are 4 issues:
- alt text is not serialized as part of the computed style
- (no-)open-quote/close-quote content is dropped from serialization
- All variants of `counters()`/`counter()` are always serialized to single-argument `counter()`
- Consecutive strings are concatenated (&quot;foo&quot; &quot;bar&quot; becomes &quot;foobar&quot;)

Also remove `CSSPrimitiveValue::createCounterName` and related code, since we now always use the more accurate `CSSCounterValue` to serialize counters.

* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-computed.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-valid.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-content/parsing/content-computed-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-content/parsing/content-valid-expected.txt: Added.
Add WPTs to avoid regressing this bug.

* LayoutTests/fast/css-generated-content/008.html:
* LayoutTests/fast/css-generated-content/013-expected.html: Added.
* LayoutTests/fast/css-generated-content/016-expected.html: Added.
* LayoutTests/fast/css-generated-content/016.html:
* LayoutTests/platform/glib/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/glib/fast/css-generated-content/016-expected.txt: Removed.
* LayoutTests/platform/gtk/fast/css-generated-content/013-expected.png: Removed.
* LayoutTests/platform/gtk/fast/css-generated-content/016-expected.png: Removed.
* LayoutTests/platform/ios-simulator/fast/css-generated-content/013-expected.png: Removed.
* LayoutTests/platform/ios/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/ios/fast/css-generated-content/016-expected.txt: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/013-expected.png: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/016-expected.png: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/016-expected.txt: Removed.
* LayoutTests/platform/wincairo/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/wincairo/fast/css-generated-content/016-expected.txt: Removed.
Make these LayoutTests more robust, now that we no longer concatenate text nodes.

* Source/WebCore/css/CSSCounterValue.cpp:
(WebCore::CSSCounterValue::customCSSText const): Fix customCSSText to also detect the customIdent version of decimal.
* Source/WebCore/css/CSSPrimitiveValue.cpp: Remove CSS_COUNTER_NAME type
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::addDerivedHash const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
(WebCore::CSSPrimitiveValue::createCounterName): Deleted.
* Source/WebCore/css/CSSPrimitiveValue.h: Remove CSS_COUNTER_NAME type
* Source/WebCore/css/CSSPrimitiveValueMappings.h: Add conversion for QuoteType
* Source/WebCore/css/CSSUnits.cpp: Remove CSS_COUNTER_NAME type
(WebCore::unitCategory):
(WebCore::operator&lt;&lt;):
* Source/WebCore/css/CSSUnits.h: Remove CSS_COUNTER_NAME type
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::contentToCSSValue): Handle quotes and fix handling of counters
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::hasDoubleValue): Remove CSS_COUNTER_NAME type
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setContent): Stop concatenating consecutive strings

Canonical link: <a href="https://commits.webkit.org/272476@main">https://commits.webkit.org/272476@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9598daf40d73fc66b5b1c421b541653bb32f5014

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31785 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10475 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34285 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/28784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32578 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12835 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7718 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32146 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/8836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/28381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/7628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/7797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/28294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35632 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/28900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/28744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/7886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/5885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/31767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/9541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/28101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/8558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4150 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/8408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSCounterValue.cpp:
```diff
@@ -53,8 +53,9 @@ bool CSSCounterValue::equals(const CSSCounterValue& other) const
 {
-    auto listStyleSeparator = m_counterStyle->valueID() == CSSValueDecimal ? ""_s : ", "_s;
-    auto listStyleLiteral = m_counterStyle->valueID() == CSSValueDecimal ? ""_s : counterStyleCSSText();
+    bool isDecimal = m_counterStyle->valueID() == CSSValueDecimal || (m_counterStyle->isCustomIdent() && m_counterStyle->customIdent() == "decimal"_s);
```
Should this be ASCII case insensitive? Typically I like to add a test case for that just to confirm. Why isn't the custom value parsed in some way? Seems strange to have this explicit check needed only in serialization without shared code since you'd think we'd need to code that uses this to render would need the same logic.

nt1m replied:
I added a testcase, and we don't need a case-insensitive comparison here. I presume we have some code somewhere that lowercases custom-idents.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2102,19 +2103,28 @@ static Ref<CSSValue> altTextToCSSValue(const RenderStyle& style)
-            list.append(CSSPrimitiveValue::createCounterName(counterContentData->counter().identifier()));
-        else if (auto* imageContentData = dynamicDowncast<ImageContentData>(*contentData))
+        if (auto* counterContentData = dynamicDowncast<CounterContentData>(*contentData)) {
+            RefPtr counterStyle = CSSPrimitiveValue::createCustomIdent(counterContentData->counter().listStyleType().identifier);
```
`auto`?

nt1m replied:
`auto` will give `Ref` instead of `RefPtr` which is what is expected here.

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2237,15 +2237,9 @@ void RenderStyle::setContent(RefPtr<StyleImage>&& image, bool add)
-        }
-    } else {
+    if (add && data.content)
+        lastContent(*data.content).setNext(makeUnique<TextContentData>(string));
```
Is there no room for an optimization where we concatenate? I understand that the old code didn't separate appropriately.

Or maybe it's not an effective optimization or impossible to do correctly.

nt1m replied:
We could optimize this by storing concatenated data as a separate list, but I don't think it's worth doing because `content` data is typically very short.

darinadler reviewed (approved):
Looks great.

nt1m committed:
[css-content] Fix issues with CSS `content` computed value serialization
https://bugs.webkit.org/show_bug.cgi?id=266828
rdar://120061551

Reviewed by Darin Adler.

There are 4 issues:
- alt text is not serialized as part of the computed style
- (no-)open-quote/close-quote content is dropped from serialization
- All variants of `counters()`/`counter()` are always serialized to single-argument `counter()`
- Consecutive strings are concatenated ("foo" "bar" becomes "foobar")

Also remove `CSSPrimitiveValue::createCounterName` and related code, since we now always use the more accurate `CSSCounterValue` to serialize counters.

* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-computed.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-content/parsing/content-valid.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-content/parsing/content-computed-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-content/parsing/content-valid-expected.txt: Added.
Add WPTs to avoid regressing this bug.

* LayoutTests/fast/css-generated-content/008.html:
* LayoutTests/fast/css-generated-content/013-expected.html: Added.
* LayoutTests/fast/css-generated-content/016-expected.html: Added.
* LayoutTests/fast/css-generated-content/016.html:
* LayoutTests/platform/glib/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/glib/fast/css-generated-content/016-expected.txt: Removed.
* LayoutTests/platform/gtk/fast/css-generated-content/013-expected.png: Removed.
* LayoutTests/platform/gtk/fast/css-generated-content/016-expected.png: Removed.
* LayoutTests/platform/ios-simulator/fast/css-generated-content/013-expected.png: Removed.
* LayoutTests/platform/ios/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/ios/fast/css-generated-content/016-expected.txt: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/013-expected.png: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/016-expected.png: Removed.
* LayoutTests/platform/mac/fast/css-generated-content/016-expected.txt: Removed.
* LayoutTests/platform/wincairo/fast/css-generated-content/013-expected.txt: Removed.
* LayoutTests/platform/wincairo/fast/css-generated-content/016-expected.txt: Removed.
Make these LayoutTests more robust, now that we no longer concatenate text nodes.

* Source/WebCore/css/CSSCounterValue.cpp:
(WebCore::CSSCounterValue::customCSSText const): Fix customCSSText to also detect the customIdent version of decimal.
* Source/WebCore/css/CSSPrimitiveValue.cpp: Remove CSS_COUNTER_NAME type
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::addDerivedHash const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
(WebCore::CSSPrimitiveValue::createCounterName): Deleted.
* Source/WebCore/css/CSSPrimitiveValue.h: Remove CSS_COUNTER_NAME type
* Source/WebCore/css/CSSPrimitiveValueMappings.h: Add conversion for QuoteType
* Source/WebCore/css/CSSUnits.cpp: Remove CSS_COUNTER_NAME type
(WebCore::unitCategory):
(WebCore::operator<<):
* Source/WebCore/css/CSSUnits.h: Remove CSS_COUNTER_NAME type
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::contentToCSSValue): Handle quotes and fix handling of counters
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::hasDoubleValue): Remove CSS_COUNTER_NAME type
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setContent): Stop concatenating consecutive strings

Canonical link: https://commits.webkit.org/272476@main
